Update reader.py to v0.3

constantinpape opened the pull request:
Changes for upcoming multiscale spec v0.3

constantinpape commented:
I have also removed the cv2 dependency here now, see discussion in #90.
However, I think it makes sense to wait with this PR on #80 (and also on #91).
@joshmoore do you have an ETA on #80?

joshmoore commented:
re: https://github.com/ome/ome-zarr-py/pull/89/commits/36e37cfd9b5dac9d92a1f9b6c72808b9d4909c0f cc: @will-moore and @erindiel

joshmoore commented:
> do you have an ETA on #80?

@constantinpape : I'll re-check it now, but I assume it can go in. It may not be 100%, but having a third class to compare to may actually _help_ get the API right.

constantinpape commented:
> > 
> 
> @constantinpape : I'll re-check it now, but I assume it can go in. It may not be 100%, but having a third class to compare to may actually _help_ get the API right.

Ok. I think it's best if you just merge #80 when you think it should work.
I will then adapt this PR to the changes and we can discuss any potential issues and potentially do changes here.

joshmoore commented on ome_zarr/reader.py:
```diff
@@ -265,7 +265,9 @@ def __init__(self, node: Node) -> None:
-            version = multiscales[0].get("version", "0.1")  # should this be matched with Format.version?
+            version = multiscales[0].get(
+                "version", "0.1"
+            )  # should this be matched with Format.version?
```
I took a bit of a look yesterday. One possibility would be for this whole block to move into the `Format` hierarchy.

constantinpape replied:
In general I agree, but for me the relation between the classes deriving from `Spec`, e.g. `Multiscales`, and the `Format` classes is not quite clear. I think we would need a longer zoom session to go through the library design to discuss this, so I would suggest to not tackle this in the current PR and leave this part as is for now.

constantinpape commented:
@joshmoore I finally have some time again to look into this.
The following things are left to do for this PR:
- [ ] Fix the precommit; I had a quick look, but I don't know what's going wrong there.
- [ ] Add tests for older format versions (multiscales 0.1, 0.2)
- [ ] Fix issues for 0.1; I can see that it does not work properly when checking visually with https://github.com/ome/napari-ome-zarr/pull/8. Might be related to issues with nested vs. flat directory store.

joshmoore commented on ome_zarr/scale.py:
```diff
@@ -123,19 +127,21 @@ def __create_group(
-            dsize=(sizeY // self.downscale, sizeX // self.downscale),
-            interpolation=cv2.INTER_NEAREST,
-        )
+            output_shape=(sizeY // self.downscale, sizeX // self.downscale),
```
@will-moore : can you check that this matches your cv2 use cases?

will-moore replied:
I think the output_shape needs to be `(sizeX, sizeY)` as in https://github.com/ome/ome-zarr-py/pull/71/files#diff-47d7162e6c0191b1b2d4a31da6e1445085655c7a1e63354b488b1d53ec54d1a2R421
See examples at https://www.tutorialkart.com/opencv/python/opencv-python-resize-image/ with `dim = (width, height)`

constantinpape replied:
Ok, I will double check this and see if we can cover it by some test.

constantinpape replied:
I have added tests for the scaler in e07194d. Judging from the results the current implementation is correct.

joshmoore commented on ome_zarr/format.py:
```diff
@@ -124,4 +125,15 @@ def init_store(self, path: str, mode: str = "r") -> FSStore:
+
+    @property
+    def version(self) -> str:
+        return "0.3"
```
Is there _no_ v03 specific logic that can be encapsulated here?

constantinpape replied:
I am not quite sure what you mean by this.

joshmoore replied:
Perhaps as with https://github.com/ome/ome-zarr-py/pull/89/files/41a6c6a95446014c6ea409df6fb823cd6a39f8bb#r665674289, this is roughly regarding what the `Format` class hierarchy should be responsible for, but I don't think it's a blocker, it just surprised me that there was no logic in the class.

constantinpape replied:
Ok, I think this is a larger design question. I suggest we move ahead without touching on this now.

will-moore commented:
@constantinpape with https://github.com/ome/napari-ome-zarr/pull/9 I can use this PR to open v0.3 images in `napari` (tested with 1 image), but this expects the `axes` in the metadata, so it needs a small addition to your PR:

e.g. this works:
```
diff --git a/ome_zarr/reader.py b/ome_zarr/reader.py
index 0d4019d..d9f5043 100644
--- a/ome_zarr/reader.py
+++ b/ome_zarr/reader.py
@@ -273,6 +273,7 @@ class Multiscales(Spec):
             axes = tuple(multiscales[0].get("axes", ["t", "c", "z", "y", "x"]))
             if len(set(axes) - axes_values) > 0:
                 raise RuntimeError(f"Invalid axes names: {set(axes) - axes_values}")
+            node.metadata['axes'] = axes
             datasets = [d["path"] for d in datasets]
             self.datasets: List[str] = datasets
             LOGGER.info("datasets %s", datasets)
```

will-moore commented on ome_zarr/reader.py:
```diff
@@ -262,10 +262,17 @@ def matches(zarr: ZarrLocation) -> bool:
+            # axes field was introduced in 0.3, before all data was 5d
+            axes = tuple(multiscales[0].get("axes", ["t", "c", "z", "y", "x"]))
+            if len(set(axes) - axes_values) > 0:
+                raise RuntimeError(f"Invalid axes names: {set(axes) - axes_values}")
```
As suggested in https://github.com/ome/ome-zarr-py/pull/89#issuecomment-879789276, passing axes data to `napari-ome-zarr`

```suggestion
                raise RuntimeError(f"Invalid axes names: {set(axes) - axes_values}")
            node.metadata['axes'] = axes
```

joshmoore replied:
Isn't that metadata already in the node, @will-moore? Is this just a case of needing/wanting an accessor?

will-moore replied:
Can I access it from there as in https://github.com/ome/napari-ome-zarr/pull/9/files ?

joshmoore replied:
I would have _thought_ so. What all is in `metadata.keys()` at the moment?

will-moore replied:
In `napari-ome-zarr`, the `node.metadata` (without the suggested change above) is:
```
metadata dict_keys(['name', 'visible', 'contrast_limits', 'colormap'])
```

constantinpape replied:
When I add this line I get the following issue in napari:
```
  File "/home/pape/Work/software/conda/miniconda3/envs/main/lib/python3.7/site-packages/napari/components/viewer_model.py", line 1003, in _add_layer_from_data
    layer = add_method(data, **(meta or {}))
TypeError: add_image() got an unexpected keyword argument 'axes'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/pape/Work/software/conda/miniconda3/envs/main/bin/napari", line 10, in <module>
    sys.exit(main())
  File "/home/pape/Work/software/conda/miniconda3/envs/main/lib/python3.7/site-packages/napari/__main__.py", line 420, in main
    _run()
  File "/home/pape/Work/software/conda/miniconda3/envs/main/lib/python3.7/site-packages/napari/__main__.py", line 309, in _run
    **kwargs,
  File "/home/pape/Work/software/conda/miniconda3/envs/main/lib/python3.7/site-packages/napari/view_layers.py", line 169, in view_path
    return _make_viewer_then('open', args, kwargs)
  File "/home/pape/Work/software/conda/miniconda3/envs/main/lib/python3.7/site-packages/napari/view_layers.py", line 119, in _make_viewer_then
    method(*args, **kwargs)
  File "/home/pape/Work/software/conda/miniconda3/envs/main/lib/python3.7/site-packages/napari/components/viewer_model.py", line 857, in open
    _path, kwargs, plugin=plugin, layer_type=layer_type
  File "/home/pape/Work/software/conda/miniconda3/envs/main/lib/python3.7/site-packages/napari/components/viewer_model.py", line 929, in _add_layers_with_plugins
    added.extend(self._add_layer_from_data(*_data))
  File "/home/pape/Work/software/conda/miniconda3/envs/main/lib/python3.7/site-packages/napari/components/viewer_model.py", line 1015, in _add_layer_from_data
    ) from exc
TypeError: _add_layer_from_data received an unexpected keyword argument ('axes') for layer type image
```

joshmoore replied:
with or without https://github.com/ome/napari-ome-zarr/pull/9 ? (If without, then it would be interesting to know whether or not we can be more [liberal](https://devopedia.org/postel-s-law) in accepting old and new inputs.

constantinpape replied:
Ah, sorry never mind; I tried with the wrong branch.

constantinpape replied:
> with or without [ome/napari-ome-zarr#9](https://github.com/ome/napari-ome-zarr/pull/9)

Yes, I have tried with https://github.com/ome/napari-ome-zarr/pull/8 instead and there it fails with the error from above.
I haven't tried with nr 9 yet.

constantinpape replied:
I get the same issue with nr 9.

constantinpape replied:
Works now.

joshmoore commented:
Hi @constantinpape, do you have a feeling when you want to try to get back to this PR? We have a hand-full that are all interlinked (below) and can start getting `pre-` versions ready, but just want to time things right.

 * https://github.com/ome/omero-cli-zarr/pull/70
 * https://github.com/hms-dbmi/vizarr/pull/104
 * https://github.com/ome/napari-ome-zarr/pull/9

constantinpape commented:
Hi @joshmoore,
yes I will try to get back to this next week.
The last weeks were more busy with modelzoo related things than expected so I didn't have time for this; but I should be free to work on ome.zarr related things on Monday.

joshmoore commented:
Sounds good. Regarding:

> * Fix the precommit; I had a quick look, but I don't know what's going wrong there.

https://github.com/ome/ome-zarr-py/pull/71/commits/561df25dc77e758864bab94470213c8cc8df8166 might fix your issues.

will-moore commented:
I exported some sample images from IDR using https://github.com/ome/omero-cli-zarr/pull/70 branch:

 - 2D (YX): https://minio-dev.openmicroscopy.org/idr/v0.3/idr0094-ellinger-sarscov2/10503791.zarr
 - 3D (CXY): https://minio-dev.openmicroscopy.org/idr/v0.3/idr0077-valuchova-flowerlightsheet/9836842.zarr
 - 4D (TCYX) https://minio-dev.openmicroscopy.org/idr/v0.3/idr0002-heriche-condensation/179758.zarr
 - 4D (CZYX): https://minio-dev.openmicroscopy.org/idr/v0.3/idr0062-blin-nuclearsegmentation/6001240.zarr
 - 4D (TCYX): https://minio-dev.openmicroscopy.org/idr/v0.3/idr0077-valuchova-flowerlightsheet/9836849.zarr

constantinpape commented:
@joshmoore @will-moore I have continued working on this now; the current status is:
- I have added tests for the scaler, which (I think) show that the current implementation of `nearest` is correct
    - this test is failing for `zoom`, `gaussian` and `laplacian`. Given that I have not changed anything related to these functions I assume they never worked and to mark these tests as expected failure and work on resolving this in a later PR
- I have updated `test_writer` to include older format versions and also check that the data on scale level 0 is read correctly
- I have added a test for `ome_zarr info` and s3. This test fails for this branch, but it does work on main; so it seems like some changes in this PR have broken this functionality. I did not have a closer look into the CLI yet, I am not so sure what causes this issue.
- The pre-commit is still failing.

joshmoore commented:
> * The pre-commit is still failing.

`tests/test_cli.py:22:89: E501 line too long (115 > 88 characters)`

constantinpape commented:
> tests/test_cli.py:22:89: E501 line too long (115 > 88 characters)

Ok, I thought this was more clever and would auto fix these issues ;).
Should be fixed manually now.

constantinpape commented on ome_zarr/reader.py:
```diff
@@ -338,7 +338,7 @@ def __init__(self, node: Node) -> None:
                 color = ch.get("color", None)
                 if color is not None:
-                    rgb = [(int(color[i: i + 2], 16) / 255) for i in range(0, 6, 2)]
+                    rgb = [(int(color[i : i + 2], 16) / 255) for i in range(0, 6, 2)]
```
This is a non PEP8 compliant change ... at least according to my flake8 linter.

constantinpape commented on tests/test_cli.py:
```diff
@@ -19,7 +19,7 @@ def initdir(self, tmpdir):
         urls = {
             "0.1": "https://s3.embassy.ebi.ac.uk/idr/zarr/v0.1/6001240.zarr",
-            "0.3": "https://minio-dev.openmicroscopy.org/idr/v0.3/idr0077-valuchova-flowerlightsheet/9836842.zarr"
+            "0.3": "https://minio-dev.openmicroscopy.org/idr/v0.3/idr0077-valuchova-flowerlightsheet/9836842.zarr",
```
I am not sure what's going on here; for some reason this undoes the changes in my previous commit:
https://github.com/ome/ome-zarr-py/pull/89/commits/a6ea341cdc27d7afbc82b7024004630e1461e072

constantinpape reviewed (commented):
It seems to me like the different linters in the precommits are inconsistent; I would ignore these failures for now...

constantinpape commented:
I found another issue with the CLI: files in format v0.3 are downloaded with flat instead of nested chunks:
```
$ ome_zarr download https://minio-dev.openmicroscopy.org/idr/v0.3/idr0077-valuchova-flowerlightsheet/9836842.zarr
downloading...
   9836842.zarr
to .
[########################################] | 100% Completed |  0.3s
[########################################] | 100% Completed |  0.2s
[########################################] | 100% Completed |  0.2s
[########################################] | 100% Completed |  0.6s
[########################################] | 100% Completed |  0.9s
[########################################] | 100% Completed |  1.8s
$ ls 9836842.zarr/0/
0.0.0  1.0.0  2.0.0  3.0.0
```

constantinpape commented:
@will-moore @joshmoore I found the issue with `ome_zarr info`; it was just a typo.

The problem with `ome_zarr download` seems to come when the zarr arrays are serialized with `data.to_zarr` [here](https://github.com/ome/ome-zarr-py/blob/master/ome_zarr/utils.py#L98). So this looks like a dask or zarr-python issue, which might be solved when the issues with the zarr dimension-separator are fixed.

I think this PR is good to be merged, pending minor TODOs:
- Mark failing tests as expected failures
- Can I ignore the pre-commit failures? If not I would need some help in fixing this as I don't understand what's going on there.

Let me know what you think.

constantinpape commented:
> +1 If you can get a state that's green as you expect things to work, then we can deal with everything else downstream.

Ok, it's almost there; the only thing that's sometimes failing is the test for `ome_zarr info`, I assume due to too many requests with all the separate test runs. I would suggest to ignore this, as we can decrease the number of test runs significantly once vispy (and in consequence PyQT / PySide) is removed from the dependencies.

constantinpape commented:
All green except the pre-commit :).

joshmoore commented on tests/test_cli.py:
```diff
@@ -14,6 +14,18 @@ class TestCli:
+        urls = {
+            "0.1": "https://s3.embassy.ebi.ac.uk/idr/zarr/v0.1/6001240.zarr",
+            "0.3": (
+                "https://minio-dev.openmicroscopy.org/idr/v0.3/"
```
@will-moore : are you ready for us to move some of the data to the IDR bucket?

joshmoore replied:
Migrated the flower to idr/zarr/v0.3:
![Screen Shot 2021-08-25 at 15 26 39](https://user-images.githubusercontent.com/88113/130798979-68aed3d6-e839-4b7c-9525-bf3435773cc0.png)

joshmoore commented on tests/test_writer.py:
```diff
@@ -33,14 +34,33 @@ def scaler(self, request):
+            pytest.param(
+                FormatV01,
+                id="V01",
+                marks=pytest.mark.xfail(reason="issues with dimension_separator"),
```
Recorded as https://github.com/ome/ome-zarr-py/issues/104

constantinpape commented:
This looks good now! @joshmoore do you want to draft releases here and in napari-ome-zarr before we merge this one?

constantinpape commented:
Just fyi, I am on vacations for the next two weeks, feel free to merge whenever this fits or change things if necessary. And looking forward to working v0.3 python libraries when I am back ;).

joshmoore commented:
Pushed a commit to test zarr 2.9.x. Otherwise, I think this is ready to go but just in case there are any last thoughts, now is the time.

joshmoore commented:
Moving forward with 0.0.22.
